nab/detectors/htmcore/htmcore_detector.py

```diff
@@ -0,0 +1,268 @@
+# ----------------------------------------------------------------------
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# Copyright (C) 2019, @breznak
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import math
+import datetime
+
+# htm.core imports
+from htm.bindings.sdr import SDR, Metrics
+from htm.encoders.rdse import RDSE, RDSE_Parameters
+from htm.encoders.date import DateEncoder
+from htm.bindings.algorithms import SpatialPooler
+from htm.bindings.algorithms import TemporalMemory
+from htm.algorithms.anomaly_likelihood import AnomalyLikelihood
+from htm.bindings.algorithms import Predictor
+
+from nab.detectors.base import AnomalyDetector
+
+# Fraction outside of the range of values seen so far that will be considered
+# a spatial anomaly regardless of the anomaly likelihood calculation. This
+# accounts for the human labelling bias for spatial values larger than what
+# has been seen so far.
+SPATIAL_TOLERANCE = 0.05
+
+## parameters to initialize our HTM model (encoder, SP, TM, AnomalyLikelihood)
+# TODO optional: optimize these params, either manually and/or swarming. But first keep comparable to numenta_detector
```

Comment:
TODO optimize params taken from hotgym

Reply:
The parameters for the original NAB are written down, somewhere in the original `nupic`.

Reply:
> parameters for the original NAB

yes, I've created params from the orig nab numenta_detector. But we do not score as well (localAreaDensity instead of removed numActiveColsPerInhArea, more importantly removed BacktrackingTM, which had best results on NAB,...)

Reply:
Ok, ive started looking at and playing with these parameters, and wow it runs fast!  I am optimizing a few of the parameters by hand and will push the results soon. Then i can see about hooking this up to the optimization framework

Reply:
>  I am optimizing a few of the parameters by hand and will push the results soon

please do. I've been trying to do some hand-optimizations, but I was still getting much worse results than Numenta's HTM :/

Reply:
Here is the new current score.  It's just a few points less than the target.
+   "reward_low_FN_rate": 66.1922106936328,
+    "reward_low_FP_rate": 58.7930712907694,
+    "standard": 63.081419488725054

Reply:
> new current score. It's just a few points less than the target.

that does look good! :+1: Let's merge this so we can go from there..